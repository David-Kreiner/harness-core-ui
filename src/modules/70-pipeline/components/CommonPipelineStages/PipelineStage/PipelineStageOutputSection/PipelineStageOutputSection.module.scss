/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.output-variables-container {
  margin-top: var(--spacing-medium);

  .table-row {
    display: grid;
    grid-template-columns: 3fr 3fr 1fr;
    column-gap: var(--spacing-8);
    margin-bottom: var(--spacing-medium);
    align-items: baseline;

    & :global(.bp3-form-group) {
      margin: 0;
    }

    .value-column {
      display: grid;
      &[data-type='RUNTIME'] {
        grid-template-columns: 3fr 45px;
      }
    }

    .actionButton {
      display: flex;
      justify-content: center;
    }

    &.header-row {
      padding-bottom: var(--spacing-medium);
      border-bottom: 1px solid var(--grey-100);
      font-style: normal;
      font-weight: 600;
      font-size: var(--font-size-xsmall);
      letter-spacing: 0.2px;
      color: var(--grey-600);
      text-transform: uppercase;
    }
  }
}

.addOutput {
  padding: 0px !important;
}

.tooltip {
  display: block;
  padding: var(--spacing-small);
}
