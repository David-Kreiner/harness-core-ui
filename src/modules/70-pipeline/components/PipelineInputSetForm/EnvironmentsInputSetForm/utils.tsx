/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get, isEmpty } from 'lodash-es'

import { RUNTIME_INPUT_VALUE, SelectOption } from '@harness/uicore'

import { DeploymentStageConfig } from 'services/cd-ng'
import { StageElementWrapperConfig } from 'services/pipeline-ng'

import { isValueRuntimeInput } from '@common/utils/utils'

import { StageType } from '@pipeline/utils/stageHelpers'
import { DeploymentStageElementConfig } from '@pipeline/utils/pipelineTypes'

export interface PropagateSelectOption extends SelectOption {
  infraLabel?: string
}

export function getPropagateStageOptions(
  stages: StageElementWrapperConfig[],
  stageIndex: number,
  templateTypes: {
    [key: string]: string
  }
): PropagateSelectOption[] {
  if (stages.length && stageIndex > 0) {
    const getStagesAllowedforPropagate = (stageItem: StageElementWrapperConfig): boolean => {
      if (stageItem.stage?.template) {
        const isAlreadyPropagatingValue = !!(stageItem?.stage?.template?.templateInputs?.spec as DeploymentStageConfig)
          ?.environment?.useFromStage?.stage
        const stageType = get(templateTypes, stageItem.stage.template.templateRef)
        return stageType === StageType.DEPLOY && !isAlreadyPropagatingValue
      } else {
        const isAlreadyPropagatingValue = !!(stageItem?.stage?.spec as DeploymentStageConfig)?.environment?.useFromStage
          ?.stage
        const stageType = stageItem?.stage?.type
        const isSingleEnvEmpty = isEmpty(
          (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment?.environmentRef
        )
        return stageType === StageType.DEPLOY && !isSingleEnvEmpty && !isAlreadyPropagatingValue
      }
    }

    const stageWithEnvV2 = stages.slice(0, stageIndex).filter(getStagesAllowedforPropagate)
    return stageWithEnvV2.map(stageItem => {
      if (stageItem.stage?.template) {
        return {
          label: `Stage [${stageItem.stage?.name}] - Template [${stageItem.stage.template.templateRef}]`,
          value: stageItem.stage?.identifier as string,
          infraLabel: `Stage [${stageItem.stage?.name}] - Template [${stageItem.stage.template.templateRef}]`
        }
      } else {
        const singleEnvironmentRef = (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment
          ?.environmentRef
        const environmentLabel = `Environment [${singleEnvironmentRef}]`

        const singleInfrastructureRef = isValueRuntimeInput(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment?.infrastructureDefinitions as any
        )
          ? RUNTIME_INPUT_VALUE
          : (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment?.infrastructureDefinitions?.[0]
              ?.identifier
        const infrastructureLabel = `Infrastructure [${singleInfrastructureRef}]`

        return {
          label: `Stage [${stageItem.stage?.name}] - ${environmentLabel}`,
          value: stageItem.stage?.identifier as string,
          infraLabel: `Stage [${stageItem.stage?.name}] - ${infrastructureLabel}`
        }
      }
    })
  }
  return []
}
