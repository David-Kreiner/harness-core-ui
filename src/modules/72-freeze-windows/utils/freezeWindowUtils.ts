/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FreezeDetailedResponse } from 'services/cd-ng'

export const RECURRENCE = ['Daily', 'Weekly', 'Monthly', 'Annually'] as const
export const DOES_NOT_REPEAT = 'Does not repeat'

type Scope = Exclude<FreezeDetailedResponse['freezeScope'], 'unknown' | undefined>

export const scopeText: Record<Scope, string> = {
  account: 'Account',
  org: 'Organization',
  project: 'Project'
}
