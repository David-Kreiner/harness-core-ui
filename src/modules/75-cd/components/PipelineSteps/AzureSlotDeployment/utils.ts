import { defaultTo } from 'lodash-es'
import { RUNTIME_INPUT_VALUE, MultiTypeInputType } from '@harness/uicore'
import type { StageElementWrapper } from '@pipeline/utils/pipelineTypes'
import type { StageElementConfig, StageElementWrapperConfig, PipelineInfoConfig } from 'services/pipeline-ng'
import type { InfraStructureDefinitionYaml, DeploymentStageConfig, EnvironmentYamlV2 } from 'services/cd-ng'

interface EnvironmentYamlV2Runtime extends Omit<EnvironmentYamlV2, 'infrastructureDefinitions'> {
  infrastructureDefinitions?: InfraStructureDefinitionYaml[] | string
}

interface DeploymentStageConfigAzureWebName extends Omit<DeploymentStageConfig, 'environment'> {
  environment: EnvironmentYamlV2Runtime
}

export interface DeploymentStageElementConfigAzureWebName extends Omit<StageElementConfig, 'spec'> {
  spec?: DeploymentStageConfigAzureWebName
}
export type SelectedStageType = StageElementWrapper<DeploymentStageElementConfigAzureWebName>

export function getStage(stageId: string, pipeline?: PipelineInfoConfig): StageElementWrapperConfig | undefined {
  if (pipeline?.stages) {
    let responseStage: StageElementWrapperConfig | undefined = undefined
    pipeline.stages.forEach(item => {
      if (item.stage && item.stage.identifier === stageId) {
        responseStage = item
      } else if (item.parallel) {
        return item.parallel.forEach(node => {
          if (node.stage?.identifier === stageId) {
            responseStage = node
          }
        })
      }
    })
    return responseStage
  }
  return
}

export const isRuntimeEnvId = (selectedStage: SelectedStageType): boolean => {
  const envId = selectedStage?.stage?.spec?.environment?.environmentRef
  return !!(envId && envId === RUNTIME_INPUT_VALUE)
}

export const isRuntimeInfraId = (selectedStage: SelectedStageType): boolean => {
  const infraId = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions
  return infraId === RUNTIME_INPUT_VALUE
}

export const isMultiEnv = (selectedStage: SelectedStageType): boolean => {
  const isEnvs = !!selectedStage?.stage?.spec?.environments || !!selectedStage?.stage?.spec?.environmentGroup
  return !!isEnvs
}

export const getEnvId = (selectedStage: SelectedStageType): string => {
  if (isMultiEnv(selectedStage || isRuntimeEnvId(selectedStage))) {
    return ''
  }

  return defaultTo(selectedStage?.stage?.spec?.environment?.environmentRef, '')
}

export const getInfraId = (selectedStage: SelectedStageType): string => {
  const envId = getEnvId(selectedStage)
  if (!envId || isMultiEnv(selectedStage)) {
    return ''
  }

  if (!isRuntimeEnvId(selectedStage) && !isRuntimeInfraId(selectedStage)) {
    const [infra] = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    return infra.identifier
  }

  return ''
}
/* eslint-disable */
export const getEnvIdRuntime = (stageId: string, values: any): string => {
  const stage = getStage(stageId, values) as SelectedStageType
  const envId = stage?.stage?.spec?.environment?.environmentRef
  return envId || ''
}
/* eslint-disable */
export const getInfraIdRuntime = (stageId: string, values: any): string => {
  const stage = getStage(stageId, values) as SelectedStageType

  if (stage?.stage?.spec?.environment?.infrastructureDefinitions?.[0]) {
    const [{ identifier }] = stage?.stage?.spec?.environment
      ?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    return identifier
  }
  return ''
}

export const getAllowableTypes = (selectedStage: SelectedStageType): MultiTypeInputType[] => {
  if (isMultiEnv(selectedStage)) {
    return [MultiTypeInputType.EXPRESSION]
  }

  return [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION]
}
