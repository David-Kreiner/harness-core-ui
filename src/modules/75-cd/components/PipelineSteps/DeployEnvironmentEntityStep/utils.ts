/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getMultiTypeFromValue, MultiTypeInputType, RUNTIME_INPUT_VALUE } from '@harness/uicore'
import { defaultTo, get, isEmpty, isNil, set } from 'lodash-es'
import type { EnvironmentYamlV2 } from 'services/cd-ng'
import type { DeployEnvironmentEntityConfig, DeployEnvironmentEntityFormState } from './types'

export function processInitialValues(initialValues: DeployEnvironmentEntityConfig): DeployEnvironmentEntityFormState {
  const formState = {}

  if (initialValues.environment && initialValues.environment.environmentRef) {
    if (getMultiTypeFromValue(initialValues.environment.environmentRef) === MultiTypeInputType.RUNTIME) {
      return {
        environment: RUNTIME_INPUT_VALUE
      }
    }

    const infrastructureDefinition = initialValues.environment.infrastructureDefinitions?.[0]
    const isInfrastructureRuntime =
      (initialValues.environment.infrastructureDefinitions as unknown as string) === RUNTIME_INPUT_VALUE

    return {
      environment: initialValues.environment.environmentRef,
      environmentInputs:
        getMultiTypeFromValue(initialValues.environment.environmentRef) === MultiTypeInputType.FIXED
          ? { [initialValues.environment.environmentRef]: initialValues.environment?.environmentInputs }
          : {},
      infrastructure: isInfrastructureRuntime ? RUNTIME_INPUT_VALUE : infrastructureDefinition?.identifier,
      // ? Can there be a better condition to identify?
      ...(!!infrastructureDefinition?.identifier && {
        infrastructureInputs: !isEmpty(infrastructureDefinition?.identifier)
          ? {
              [initialValues.environment.environmentRef]: {
                [infrastructureDefinition.identifier]: infrastructureDefinition?.inputs
              }
            }
          : {}
      })
    }
  } else if (initialValues.environments) {
    if (getMultiTypeFromValue(initialValues.environments?.values as unknown as string) !== MultiTypeInputType.FIXED) {
      set(formState, 'environments', initialValues.environments?.values)
      set(formState, 'environmentInputs', {})
      set(formState, 'infrastructures', [])
      set(formState, 'infrastructureInputs', {})
    } else {
      defaultTo(initialValues.environments?.values, []).map((environment: EnvironmentYamlV2, index: number) => {
        set(formState, `environments.${index}`, {
          label: environment.environmentRef,
          value: environment.environmentRef
        })
        set(formState, `environmentInputs.${environment.environmentRef}`, environment.environmentInputs)
        set(formState, `serviceOverrideInputs.${environment.environmentRef}`, environment.serviceOverrideInputs)
        if (Array.isArray(environment.infrastructureDefinitions)) {
          environment.infrastructureDefinitions.map((infrastructure, infrastructureIndex) => {
            set(formState, `infrastructures.${environment.environmentRef}.${infrastructureIndex}`, {
              label: infrastructure.identifier,
              value: infrastructure.identifier
            })
            set(
              formState,
              `infrastructureInputs.${environment.environmentRef}.${infrastructure.identifier}`,
              infrastructure.inputs
            )
          })
        } else {
          set(formState, `infrastructures.${environment.environmentRef}`, environment.infrastructureDefinitions)
        }
      })
    }
    set(formState, `parallel`, defaultTo(initialValues.environments?.metadata?.parallel, true))
  } else if (initialValues.environmentGroup && initialValues.environmentGroup.envGroupRef) {
    set(formState, `environmentGroup`, initialValues.environmentGroup.envGroupRef)

    if (
      getMultiTypeFromValue(initialValues.environmentGroup?.envGroupRef as unknown as string) ===
      MultiTypeInputType.FIXED
    ) {
      if (
        getMultiTypeFromValue(initialValues.environmentGroup?.environments as unknown as string) !==
        MultiTypeInputType.FIXED
      ) {
        set(formState, 'environments', initialValues.environmentGroup.environments)
        set(formState, 'environmentInputs', {})
        set(formState, 'infrastructures', [])
        set(formState, 'infrastructureInputs', {})
      } else {
        // This section is same as multi env. Move to function
        defaultTo(initialValues.environmentGroup.environments, []).map(
          (environment: EnvironmentYamlV2, index: number) => {
            set(formState, `environments.${index}`, {
              label: environment.environmentRef,
              value: environment.environmentRef
            })
            set(formState, `environmentInputs.${environment.environmentRef}`, environment.environmentInputs)
            set(formState, `serviceOverrideInputs.${environment.environmentRef}`, environment.serviceOverrideInputs)
            if (Array.isArray(environment.infrastructureDefinitions)) {
              environment.infrastructureDefinitions.map((infrastructure, infrastructureIndex) => {
                set(formState, `infrastructures.${environment.environmentRef}.${infrastructureIndex}`, {
                  label: infrastructure.identifier,
                  value: infrastructure.identifier
                })
                set(
                  formState,
                  `infrastructureInputs.${environment.environmentRef}.${infrastructure.identifier}`,
                  infrastructure.inputs
                )
              })
            } else {
              set(formState, `infrastructures.${environment.environmentRef}`, environment.infrastructureDefinitions)
            }
          }
        )
      }
    }
  }

  return formState
}

// TODO: Add type check to optionally extends keys
export type ChangeTypeOfKeys<T, N, K extends keyof T> = Pick<T, K> &
  {
    [key in keyof T]: T[key] | N
  }

export function processFormValues(data: DeployEnvironmentEntityFormState): DeployEnvironmentEntityConfig {
  if (!isNil(data.environment)) {
    // ! Do not merge this with the other returns even if they look similar. It makes it confusing to read
    if (getMultiTypeFromValue(data.environment) === MultiTypeInputType.RUNTIME) {
      return {
        environment: {
          environmentRef: RUNTIME_INPUT_VALUE,
          deployToAll: false,
          environmentInputs: RUNTIME_INPUT_VALUE as any,
          serviceOverrideInputs: RUNTIME_INPUT_VALUE as any,
          infrastructureDefinitions: RUNTIME_INPUT_VALUE as any
        }
      }
    }

    return {
      environment: {
        environmentRef: data.environment,
        ...(!!data.environmentInputs?.[data.environment] && {
          environmentInputs: data.environmentInputs[data.environment]
        }),
        deployToAll: false,
        ...(data.environment &&
          !!data.infrastructure && {
            infrastructureDefinitions:
              getMultiTypeFromValue(data.infrastructure) === MultiTypeInputType.RUNTIME
                ? (data.infrastructure as any)
                : [
                    {
                      identifier: data.infrastructure,
                      inputs: get(data, `infrastructureInputs.${data.environment}.${data.infrastructure}`)
                    }
                  ]
          })
      }
    }
  } else if (!isNil(data.environmentGroup)) {
    if (getMultiTypeFromValue(data.environmentGroup) === MultiTypeInputType.RUNTIME) {
      return {
        environmentGroup: {
          envGroupRef: RUNTIME_INPUT_VALUE,
          deployToAll: RUNTIME_INPUT_VALUE as any,
          environments: RUNTIME_INPUT_VALUE as any
        }
      }
    }

    return {
      environmentGroup: {
        envGroupRef: data.environmentGroup,
        deployToAll:
          getMultiTypeFromValue(data.environments) === MultiTypeInputType.RUNTIME
            ? (RUNTIME_INPUT_VALUE as any)
            : isEmpty(data.environments),
        environments: Array.isArray(data.environments)
          ? data.environments?.map(environment => {
              const selectedInfrastructures = data.infrastructures?.[environment.value as string]
              const isInfrastructureRuntime = (selectedInfrastructures as unknown as string) === RUNTIME_INPUT_VALUE

              return {
                environmentRef: environment.value as string,
                ...(!!data.environmentInputs?.[environment.value as string] && {
                  environmentInputs: data.environmentInputs[environment.value as string]
                }),
                deployToAll: isInfrastructureRuntime ? (RUNTIME_INPUT_VALUE as any) : !selectedInfrastructures,
                ...(!!selectedInfrastructures && {
                  infrastructureDefinitions: Array.isArray(selectedInfrastructures)
                    ? selectedInfrastructures.map(infrastructure => ({
                        identifier: infrastructure.value as string,
                        inputs: data.infrastructureInputs?.[environment.value as string][infrastructure.value as string]
                      }))
                    : selectedInfrastructures
                })
              }
            })
          : data.environments
      }
    }
  } else if (!isNil(data.environments)) {
    return {
      environments: {
        metadata: {
          parallel: data.parallel
        },
        values: Array.isArray(data.environments)
          ? data.environments?.map(environment => {
              const selectedInfrastructures = data.infrastructures?.[environment.value as string]
              const isInfrastructureRuntime = (selectedInfrastructures as unknown as string) === RUNTIME_INPUT_VALUE

              return {
                environmentRef: environment.value as string,
                ...(!!data.environmentInputs?.[environment.value as string] && {
                  environmentInputs: data.environmentInputs[environment.value as string]
                }),
                deployToAll: isInfrastructureRuntime ? (RUNTIME_INPUT_VALUE as any) : !selectedInfrastructures,
                ...(!!selectedInfrastructures && {
                  infrastructureDefinitions: Array.isArray(selectedInfrastructures)
                    ? selectedInfrastructures.map(infrastructure => ({
                        identifier: infrastructure.value as string,
                        inputs: data.infrastructureInputs?.[environment.value as string][infrastructure.value as string]
                      }))
                    : selectedInfrastructures
                })
              }
            })
          : data.environments
      }
    }
  }
  return {}
}
