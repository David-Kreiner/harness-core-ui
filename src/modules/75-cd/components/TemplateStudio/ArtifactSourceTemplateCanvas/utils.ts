import type { ArtifactType } from '@pipeline/components/ArtifactsSelection/ArtifactInterface'
import { ArtifactToConnectorMap } from '@pipeline/components/ArtifactsSelection/ArtifactHelper'

export interface Params {
  oldConnectorRef?: string
  newConnectorRef?: string
  artifactType: ArtifactType
}

export const isConnectorRefUpdated = (params: Params) => {
  const { oldConnectorRef, newConnectorRef, artifactType } = params
  const isConnectorRefApplicable = Boolean(ArtifactToConnectorMap[artifactType])

  return Boolean(isConnectorRefApplicable && oldConnectorRef && newConnectorRef && oldConnectorRef !== newConnectorRef)
}
