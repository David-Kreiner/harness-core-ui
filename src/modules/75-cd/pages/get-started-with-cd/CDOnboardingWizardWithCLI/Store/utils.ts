export const parseJSON = (str: string): object => {
  try {
    return JSON.parse(str)
  } catch (e) {
    return {}
  }
}
