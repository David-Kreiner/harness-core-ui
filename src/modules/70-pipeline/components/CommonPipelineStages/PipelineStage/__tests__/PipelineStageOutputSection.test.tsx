/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, waitFor, screen, findByRole, getByRole, within, RenderResult, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { RUNTIME_INPUT_VALUE } from '@harness/uicore'
import type { FormikProps } from 'formik'
import { findDialogContainer, TestWrapper } from '@common/utils/testUtils'
import {
  PipelineContext,
  PipelineContextInterface
} from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import routes from '@common/RouteDefinitions'
import { accountPathProps, pipelineModuleParams, projectPathProps } from '@common/utils/routeUtils'
import { StageErrorContext } from '@pipeline/context/StageErrorContext'
import { yamlStringify } from '@common/utils/YamlHelperMethods'
import * as PipelineVariablesContext from '@pipeline/components/PipelineVariablesContext/PipelineVariablesContext'
import {
  errorContextProvider,
  getDummyPipelineContextValue,
  getMockFor_useGetPipeline,
  getModuleParams
} from './PipelineStageHelper'
import { PipelineStageOutputSection } from '../PipelineStageOutputSection/PipelineStageOutputSection'
import { VariableOutputPanel } from '../PipelineStageOutputSection/VariableOutputPanel'
import variablesPipeline from './variablesPipeline.json'
import metadataMap from './metadataMap.json'

jest.mock('services/pipeline-ng', () => ({
  useCreateVariablesV2: jest.fn(() => ({
    mutate: jest.fn(() =>
      Promise.resolve({
        data: {
          yaml: yamlStringify({ pipeline: variablesPipeline }),
          metadataMap
        }
      })
    ),
    loading: false,
    cancel: jest.fn()
  })),
  useGetPipeline: jest.fn(() => getMockFor_useGetPipeline())
}))

jest.mock('@harness/monaco-yaml/lib/esm/languageservice/yamlLanguageService', () => ({
  getLanguageService: jest.fn()
}))

jest.mock('@common/components/MonacoEditor/MonacoEditor')
jest.mock('@common/components/YAMLBuilder/YamlBuilder')
jest.mock('@common/utils/YamlUtils', () => ({}))
jest.mock('lodash-es', () => ({
  ...(jest.requireActual('lodash-es') as Record<string, any>),
  debounce: jest.fn(fn => {
    fn.cancel = jest.fn()
    return fn
  }),
  noop: jest.fn()
}))

jest.useFakeTimers({ advanceTimers: true })
jest.mock('resize-observer-polyfill', () => {
  class ResizeObserver {
    static default = ResizeObserver
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    observe() {
      // do nothing
    }
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    unobserve() {
      // do nothing
    }
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    disconnect() {
      // do nothing
    }
  }
  return ResizeObserver
})

jest.spyOn(PipelineVariablesContext, 'usePipelineVariables').mockImplementation((): any => {
  return {
    variablesPipeline: { pipeline: variablesPipeline },
    metadataMap
  }
})

const renderOutputSectionComponent = (value?: PipelineContextInterface): RenderResult => {
  const pipelineContextMockValue = getDummyPipelineContextValue()
  const pipelineContextValue = value ?? pipelineContextMockValue

  return render(
    <TestWrapper path={TEST_PATH} pathParams={getModuleParams('chainedPipeline', 'cd')}>
      <PipelineContext.Provider value={pipelineContextValue}>
        <StageErrorContext.Provider value={errorContextProvider}>
          <PipelineStageOutputSection />
        </StageErrorContext.Provider>
      </PipelineContext.Provider>
    </TestWrapper>
  )
}

const TEST_PATH = routes.toPipelines({ ...accountPathProps, ...projectPathProps, ...pipelineModuleParams })

describe('PipelineStageOutputSection Test', () => {
  test('Empty Output view', async () => {
    const { container } = renderOutputSectionComponent()
    expect(await screen.findByText('pipeline.pipelineChaining.pipelineOutputs')).toBeDefined()
    expect(screen.queryByTestId('output-row-0')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot('PipelineStageOutputSection - No Output')
  })

  test('Filled Outputs view', async () => {
    const pipelineContextMockValue = getDummyPipelineContextValue()
    renderOutputSectionComponent({
      ...pipelineContextMockValue,
      state: {
        ...pipelineContextMockValue.state,
        selectionState: {
          selectedSectionId: 'OUTPUTS',
          selectedStageId: 'parStage2',
          selectedStepId: undefined
        }
      },
      getStageFromPipeline: jest.fn(() => {
        return { stage: pipelineContextMockValue.state.pipeline.stages?.[1], parent: undefined }
      })
    } as any)

    expect(await screen.findByText('pipeline.pipelineChaining.pipelineOutputs')).toBeDefined()

    // Configure Option Test
    const configureButton = screen.getByRole('button', {
      name: /cog/i
    })
    await userEvent.click(configureButton)
    const configureOptionsDialog = findDialogContainer() as HTMLElement
    expect(
      await findByRole(configureOptionsDialog, 'heading', {
        name: 'common.configureOptions.configureOptions'
      })
    )
    userEvent.type(getByRole(configureOptionsDialog, 'textbox'), '12')
    userEvent.click(
      getByRole(configureOptionsDialog, 'button', {
        name: /submit/i
      })
    )

    // Delete button test
    const deleteOutputButton = await screen.findByTestId('delete-output-2')
    await userEvent.click(deleteOutputButton)
    expect(screen.queryByTestId('output-row-2')).not.toBeInTheDocument()

    // Add Output button test
    const newOutputAddButton = screen.getByRole('button', {
      name: 'pipeline.pipelineChaining.newOutput'
    })
    userEvent.click(newOutputAddButton)
    const thirdOutputRow = await screen.findByTestId('output-row-2')
    expect(thirdOutputRow).toBeInTheDocument()

    // Check Onchange - multi type button
    const multiTypeButton = within(thirdOutputRow).getByTestId('multi-type-button')
    userEvent.click(multiTypeButton)
    const runtimeInputText = await screen.findByText('Runtime input')
    expect(runtimeInputText).toBeInTheDocument()
    userEvent.click(runtimeInputText)
    expect(await within(thirdOutputRow).findByPlaceholderText(RUNTIME_INPUT_VALUE)).toBeInTheDocument()
  })

  test('validate duplicate outputs name', async () => {
    const pipelineContextMockValue = getDummyPipelineContextValue()
    const formikRef = React.createRef<FormikProps<unknown>>()

    render(
      <TestWrapper path={TEST_PATH} pathParams={getModuleParams('chainedPipeline', 'cd')}>
        <PipelineContext.Provider value={pipelineContextMockValue}>
          <StageErrorContext.Provider value={errorContextProvider}>
            <VariableOutputPanel formikRef={formikRef} />
          </StageErrorContext.Provider>
        </PipelineContext.Provider>
      </TestWrapper>
    )

    // Add Outputs
    const newOutputAddButton = screen.getByRole('button', {
      name: 'pipeline.pipelineChaining.newOutput'
    })
    await userEvent.click(newOutputAddButton)
    const firstOutputRow = await screen.findByTestId('output-row-0')
    const firstOutputNameInput = await within(firstOutputRow).findByPlaceholderText(
      'pipeline.pipelineChaining.enterOutputName'
    )
    await userEvent.type(firstOutputNameInput, 'output1')

    await userEvent.click(newOutputAddButton)
    const secondOutputRow = await screen.findByTestId('output-row-1')
    const secondOutputNameInput = await within(secondOutputRow).findByPlaceholderText(
      'pipeline.pipelineChaining.enterOutputName'
    )
    await userEvent.type(secondOutputNameInput, 'output1')
    act(() => {
      formikRef.current?.submitForm()
    })
    expect(await screen.findAllByText('pipeline.pipelineChaining.outputAlreadyExists')).toHaveLength(2)
    expect(formikRef.current?.errors).toEqual({
      outputs: [
        {
          name: 'pipeline.pipelineChaining.outputAlreadyExists',
          value: 'common.validation.valueIsRequired'
        },
        {
          name: 'pipeline.pipelineChaining.outputAlreadyExists',
          value: 'common.validation.valueIsRequired'
        }
      ]
    })

    userEvent.clear(secondOutputNameInput)
    userEvent.type(secondOutputNameInput, 'output2')
    await waitFor(() =>
      expect(
        within(firstOutputRow).queryByText('pipeline.pipelineChaining.outputAlreadyExists')
      ).not.toBeInTheDocument()
    )
  })

  test('Readonly view', async () => {
    const pipelineContextMockValue = getDummyPipelineContextValue()
    renderOutputSectionComponent({
      ...pipelineContextMockValue,
      state: {
        ...pipelineContextMockValue.state,
        selectionState: {
          selectedSectionId: 'OUTPUTS',
          selectedStageId: 'parStage2',
          selectedStepId: undefined
        }
      },
      getStageFromPipeline: jest.fn(() => {
        return { stage: pipelineContextMockValue.state.pipeline.stages?.[1], parent: undefined }
      }),
      isReadonly: true
    } as any)

    expect(await screen.findByDisplayValue('<+pipeline.name>')).toHaveAttribute('disabled')
    expect(screen.queryByTestId('delete-output-0')).not.toBeInTheDocument()
  })

  test('Render correct fqn and local name of chained pipeline outputs', async () => {
    const pipelineContextMockValue = getDummyPipelineContextValue()
    renderOutputSectionComponent({
      ...pipelineContextMockValue,
      state: {
        ...pipelineContextMockValue.state,
        selectionState: {
          selectedSectionId: 'OUTPUTS',
          selectedStageId: 'parStage2',
          selectedStepId: undefined
        }
      },
      getStageFromPipeline: jest.fn(() => {
        return { stage: variablesPipeline.stages?.[1], parent: undefined }
      })
    } as any)

    expect(await screen.findByText('pipeline.pipelineChaining.pipelineOutputs')).toBeDefined()
    const firstOutputRow = await screen.findByTestId('output-row-0')
    const localFQNCopyButton = firstOutputRow.querySelector('span[data-icon="copy-alt"]')
    await userEvent.click(localFQNCopyButton as HTMLElement)

    const usedWithinStageText = await screen.findByText('common.usedWithinStage')
    expect(usedWithinStageText).toBeInTheDocument()
    const localName = screen.getByText('stages.parStage2.output.output1')
    const fqn = screen.getByText('pipeline.stages.parStage2.output.output1')
    expect(localName).toBeInTheDocument()
    expect(fqn).toBeInTheDocument()
  })
})
