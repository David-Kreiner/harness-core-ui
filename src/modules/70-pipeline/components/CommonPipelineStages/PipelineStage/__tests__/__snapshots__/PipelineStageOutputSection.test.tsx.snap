// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PipelineStageOutputSection Test Empty Output view: PipelineStageOutputSection - No Output 1`] = `
<div>
  <div
    class="pipelineStageOverviewWrapper"
  >
    <div
      class="StyledProps--font StyledProps--main content"
    >
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-variation-h5 StyledProps--margin-bottom-medium"
      >
        pipeline.pipelineChaining.pipelineOutputs
      </p>
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-variation-h6 StyledProps--font-weight-light StyledProps--margin-bottom-small"
      >
        pipeline.pipelineChaining.outputTabSummaryDetail
      </p>
      <div
        class="bp3-card bp3-elevation-0 Card--card outputPanel"
        id="outputs"
      >
        <div
          class="outputVariablesContainer"
        >
          <button
            aria-label="pipeline.pipelineChaining.newOutput"
            class="bp3-button Button--button StyledProps--font StyledProps--main addOutput Button--with-current-color Button--withLeftIcon Button--variation Button--variation-link Button--size Button--small"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-plus StyledProps--main StyledProps--padding-right-xsmall"
              icon="plus"
            >
              <svg
                data-icon="plus"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  plus
                </desc>
                <path
                  d="M13 7H9V3c0-.55-.45-1-1-1s-1 .45-1 1v4H3c-.55 0-1 .45-1 1s.45 1 1 1h4v4c0 .55.45 1 1 1s1-.45 1-1V9h4c.55 0 1-.45 1-1s-.45-1-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <span
              class="bp3-button-text"
            >
              pipeline.pipelineChaining.newOutput
            </span>
          </button>
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main actionButtons StyledProps--margin-top-xxlarge"
      />
    </div>
  </div>
</div>
`;
